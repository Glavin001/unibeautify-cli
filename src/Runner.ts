import chalk from "chalk";

import {
  getSupportedLanguages,
  findInstalledBeautifiers,
  setupUnibeautify,
  getAllLanguages,
} from "./index";
import { BeautifyData } from "unibeautify";
import * as cosmiconfig from "cosmiconfig";
import * as fs from "fs";

export class Runner {
  protected stdout: NodeJS.WriteStream = process.stdout;
  protected stderr: NodeJS.WriteStream = process.stderr;

  public beautify(programArgs: IArgs): Promise<void> {
    const { language, filePath } = programArgs;
    return setupUnibeautify().then(unibeautify => {
      if (!language) {
        this.writeError("A language is required.");
        return this.exit(1);
      }
      return Promise.all([
        this.readConfig(programArgs),
        this.readText(filePath),
      ]).then(([config, text]) => {
        const data: BeautifyData = {
          filePath: filePath,
          languageName: language,
          options: (config as any) || {},
          text,
        };
        return unibeautify
          .beautify(data)
          .then((result: string) => {
            this.writeOut(result);
            return this.exit(0);
          })
          .catch((error: Error) => {
            this.writeError(error.message);
            return this.exit(1);
          });
      });
    });
  }

  private readConfig(programArgs: IArgs): Promise<any> {
    const { configFile, configJson } = programArgs;
    if (configJson) {
      return Promise.resolve(this.parseConfig(configJson));
    } else {
      return this.configFile(configFile);
    }
  }

  private async readText(filePath?: string): Promise<string> {
    if (this.isTerminal && filePath) {
      return await this.readFile(filePath);
    } else {
      return await this.readFromStdin();
    }
  }

  public support(options: {
    json?: boolean;
    languages?: boolean;
    beautifiers?: boolean;
    all?: boolean;
  }): Promise<void> {
    const printer: (info: SupportInfo) => void = options.json
      ? this.jsonPrinter
      : this.listPrinter;
    const info: SupportInfo = {};
    return setupUnibeautify().then(async unibeautify => {
      if (options.languages) {
        if (options.all) {
          info["languages"] = getAllLanguages();
        } else {
          info["languages"] = getSupportedLanguages();
        }
      }
      if (options.beautifiers) {
        const beautifiers = await findInstalledBeautifiers();
        info["beautifiers"] = beautifiers;
      }
      if (Object.keys(info).length === 0) {
        this.writeError("Nothing to show");
        this.exit(1);
      } else {
        printer(info);
        this.exit(0);
      }
    });
  }

  private parseConfig(configJson?: string): object {
    let config = {};
    if (configJson) {
      try {
        config = JSON.parse(configJson);
      } catch (e) {
        // FIXME: I think this should return an error instead of stderr
        this.writeError(e.message);
        this.exit(2);
      }
    }
    return config;
  }

  private configFile(configFile?: string, filePath?: string) {
    const configExplorer = cosmiconfig("unibeautify", {});
    const promise = configFile ? configExplorer.load(configFile) : configExplorer.search(filePath);
    return promise.then(result => (result ? result.config : null))
      .catch(error =>
        Promise.reject(new Error(`Could not load configuration file ${configFile}`))
      );
  }

  protected get isTerminal(): boolean {
    return Boolean(process.stdin.isTTY);
  }

  protected readFromStdin(): Promise<string> {
    return new Promise((resolve, reject) => {
      let text = "";
      process.stdin.resume();
      process.stdin.on("data", (data: string) => {
        text = data.toString();
      });
      process.stdin.on("end", () => {
        return resolve(text);
      });
      process.stdout.on("error", (err: any) => {
        if (err.code === "EPIPE") {
          return this.exit(1);
        }
        process.emit("warning", err);
      });
    });
  }

  private jsonPrinter = (info: SupportInfo) => {
    this.writeOut(JSON.stringify(info, null, 2));
    // unibeautify:ignore-next-line
  }

  private listPrinter = (info: SupportInfo) => {
    Object.keys(info).forEach(section => {
      this.writeOut(chalk.blue(`Supported ${section}`));
      const items = info[section];
      items.forEach((item, index) => this.writeOut(`${index + 1}. ${item}`));
    });
    // unibeautify:ignore-next-line
  }

  protected writeOut(text: string): void {
    this.stdout.write(text + "\n");
  }

  protected writeError(text: string): void {
    this.stderr.write(text + "\n");
  }

  protected exit(exitCode: number): void {
    process.exit(exitCode);
  }

  private readFile(filePath: string): Promise<string> {
    return new Promise((resolve, reject) => {
      fs.readFile(filePath, (error, data) => {
        if (error) {
          return reject(error);
        }
        return resolve(data.toString());
      });
    });
  }
}

/**
  Arguments parsed for program.
  */
export interface IArgs {
  args: string[];
  language?: string;
  outFile?: string;
  replace?: boolean;
  configFile?: string;
  configJson?: string;
  filePath?: string;
}

interface SupportInfo {
  [section: string]: string[];
}
