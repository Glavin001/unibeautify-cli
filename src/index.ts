import unibeautify, { Unibeautify } from "unibeautify";
const requireg = require("requireg");
import { spawn } from "child_process";

/**
Find all globally installed beautifiers.
*/
export function findInstalledBeautifiers(): Promise<string[]> {
  return new Promise((resolve, reject) => {
    runSpawn("npm", ["ls", "-g", "--depth=0", "--json"]).then(
      globalPackages => {
        const packageJSON = JSON.parse(globalPackages).dependencies;
        const packageNames = Object.keys(packageJSON);
        const beautifierNames = packageNames.filter((name: string) => {
          return /beautifier-.*/.test(name);
        });
        return resolve(beautifierNames);
      }
    );
  });
}

function runSpawn(exe: string, args: string[]): Promise<any> {
  return new Promise((resolve, reject) => {
    const ls = spawn(exe, args);
    let stdout = "";
    ls.stdout.on("data", data => {
      return (stdout += data);
    });
    // ls.stderr.on("data", (data) => {
    //   console.log(`stderr: ${data}`);
    // });
    ls.on("close", code => {
      return resolve(stdout);
    });
    ls.on("error", err => {
      // tslint:disable-next-line:no-console
      console.error(err);
    });
  });
}

export function loadBeautifiers(beautifierNames: string[]): Unibeautify {
  const beautifiers = beautifierNames.map(beautifierName => {
    return requireg(beautifierName).default;
  });
  return unibeautify.loadBeautifiers(beautifiers);
}

export function loadInstalledBeautifiers(): Promise<Unibeautify> {
  return findInstalledBeautifiers().then((beautifierNames: string[]) => {
    return loadBeautifiers(beautifierNames);
  });
}

export function setupUnibeautify(): Promise<Unibeautify> {
  return loadInstalledBeautifiers();
}

export function getSupportedLanguages(): string[] {
  return unibeautify.supportedLanguages.map(language => {
    return language.name;
  });
}

export function getAllLanguages(): string[] {
  return unibeautify.getLoadedLanguages().map(language => {
    return language.name;
  });
}
