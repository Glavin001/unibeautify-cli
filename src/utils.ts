import unibeautify, { Unibeautify } from "unibeautify";
const requireg = require("requireg");
const gSearch = require("g-search");

/**
Find all globally installed beautifiers.
*/
export function findInstalledBeautifiers(): Promise<string[]> {
  return new Promise((resolve, reject) => {
    gSearch()
      .then((globalPackages: GlobalSearchResult[]) => {
        const packageNames = globalPackages.map(pkg => pkg.name);
        const beautifierNames = packageNames.filter(isBeautifierPackageName);
        return resolve(beautifierNames);
      })
      .catch((error: Error) => {
        return reject(error);
      });
  });
}

export function isBeautifierPackageName(packageName: string): boolean {
  return /beautifier-.*/.test(packageName);
}

export function loadBeautifiers(beautifierNames: string[]): Unibeautify {
  const beautifiers = beautifierNames
    .map(beautifierName => {
      try {
        return requireg(beautifierName).default;
      } catch (error) {
        console.error(error);
        return null;
      }
    })
    .filter(Boolean);
  return unibeautify.loadBeautifiers(beautifiers);
}

export function loadInstalledBeautifiers(): Promise<Unibeautify> {
  return findInstalledBeautifiers().then((beautifierNames: string[]) => {
    return loadBeautifiers(beautifierNames);
  });
}

export function setupUnibeautify(): Promise<Unibeautify> {
  return loadInstalledBeautifiers();
}

export function getSupportedLanguages(): string[] {
  return unibeautify.supportedLanguages.map(language => {
    return language.name;
  });
}

export function getAllLanguages(): string[] {
  return unibeautify.getLoadedLanguages().map(language => {
    return language.name;
  });
}

export interface GlobalSearchResult {
  name: string;
  version: string;
  location: string;
}
