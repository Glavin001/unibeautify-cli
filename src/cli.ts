#!/usr/bin/env node

/**
 * Module dependencies.
 */
import * as program from "yargs";

import { Runner } from "./Runner";

program
  .version()
  .command({
    builder: {
      "config-file": {
        alias: "c",
        demandOption: false,
        describe: "Beautifier configuration file",
        type: "string",
      },
      "config-json": {
        demandOption: false,
        describe: "Beautifier configuration as stringified JSON",
        type: "string",
      },
      "file-path": {
        alias: "f",
        demandOption: false,
        describe: "Path of the file to beautify from stdin",
        type: "string",
      },
      language: {
        alias: "l",
        demandOption: true,
        describe: "Language of file to beautify",
        type: "string",
      },
      "out-file": {
        alias: "o",
        demandOption: false,
        describe: "Output file of beautified results",
        type: "string",
      },
      replace: {
        alias: "r",
        default: false,
        demandOption: false,
        describe: "Replace file(s) with beautified results",
        type: "boolean",
      },
    },
    command: "*",
    describe: "Beautify Files",
    handler: (argv) => {
      const main = new Runner();
      main.beautify(argv);
    }
  })
  .command({
    aliases: ["list", "ls"],
    builder: {
      all: {
        alias: "a",
        default: false,
        demandOption: false,
        describe: "List all languages supported",
      },
      beautifiers: {
        alias: "b",
        default: false,
        demandOption: false,
        describe: "List installed beautifiers",
      },
      json: {
        alias: "j",
        default: false,
        demandOption: false,
        describe: "List as JSON array",
      },
      languages: {
        alias: "l",
        default: false,
        demandOption: false,
        describe: "List supported languages based on installed beautifiers",
      },
    },
    command: "support",
    describe: "List languages and beautifiers",
    handler: (argv) => {
      const main = new Runner();
      main.support(argv);
    }
  })
  .help()
  .argv;


  // .option("-i, --input-file [file]", "Input file to be beautified")
  // .option("-b, --beautifiers [beautifier...]", "Installed Beautifiers to load")
  // .option("--input-dir [directory]", "Input directory of files to be beautified")
  // .option("--output-dir [directory]", "Output directory of beautified results")

// const programArgs: IArgs = program.parse(process.argv);
// const main = new Runner();
// main.beautify(programArgs);
