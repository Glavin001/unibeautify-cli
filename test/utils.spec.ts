import {findInstalledBeautifiers} from "../src/utils";

test("findInstalledBeautifiers", () => {
    expect.assertions(2);
    return findInstalledBeautifiers().then(beautifierNames => {
        expect(beautifierNames).toHaveLength(1);
        expect(beautifierNames).toMatchSnapshot();
    });
});
