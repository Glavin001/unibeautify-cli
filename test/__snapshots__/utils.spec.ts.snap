// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findInstalledBeautifiers 1`] = `
Array [
  "@unibeautify/beautifier-eslint",
]
`;
